### PR TITLE
Kafka Catch Up

### PR DESCRIPTION
**Pull request by dmdashenkov**
In this PR we introduce a utility for the `ProjectionRepository` catch up.

Based on Kafka Streams API, it allows us to perform the read side catch up with guaranteed event delivery, leaving the most (but optional) config to the framework user.

See `KafkaCatchUp`  class for the user manual.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.
Note: this PR still contains 1 `TODO` item. It is to be fixed separately. [An issue](https://github.com/SpineEventEngine/todo-list/issues/37) is created for this TODO.

File: deployment/local-kafka/src/main/java/io/spine/examples/todolist/server/ConfigFiles.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+/**
+ * A utility for working with the config files contained in the classpath.
+ *
+ * @author Dmytro Dashenkov
+ */
+final class ConfigFiles {
```
Comment (armiol) [inline deployment/local-kafka/src/main/java/io/spine/examples/todolist/server/ConfigFiles.java]: Please find a more generic name. Such as `Configurations`. In which case `loadConfig` will become `load`. And `Properties` will be facaded with some interface as well (e.g. `Configuration`, that is responsible for parsing raw values from the string-to-string contents of the raw `Properties`).

Right now it looks a bit under-thought-thru.

File: ext.gradle [hunk 2]
```diff
@@ -19,7 +19,7 @@
  */
 
 
-final def SPINE_VERSION = '0.9.64-SNAPSHOT'
+final def SPINE_VERSION = '0.9.76-SNAPSHOT'
```
Comment (armiol) [inline ext.gradle]: We have `0.10.1-SNAPSHOT`. You could use it right away, if possible.

File: deployment/local-kafka/src/main/resources/config/kafka-streams.properties [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+bootstrap.servers=localhost:4545
```
Comment (armiol) [inline deployment/local-kafka/src/main/resources/config/kafka-streams.properties]: Please have it commented. I would even have two comments for this file: one for the file purpose in general and the second one for the property and possible values (e.g.  "servers"  implies having more than one, but there is no evidence of this actually being true).

File: kafka-storage/src/main/java/io/spine/server/catchup/KafkaCatchUp.java [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+ *     <li>All the events are grouped by the target {@code ProjectionRepository} and dispatched
+ *     into the repository synchronously.
+ *     <li>The repository handles the events by its own.
+ * </ol>
+ *
+ * @author Dmytro Dashenkov
+ */
+public final class KafkaCatchUp {
+
+    private static final long WINDOW_SIZE_MS = 2000;
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/catchup/KafkaCatchUp.java]: Please have both properties commented. It is important as this is an example we give to people, that generally speaking aren't able to ask the dev team for casual clarifications.

File: kafka-storage/src/main/java/io/spine/server/catchup/KafkaCatchUp.java [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+
+    /**
+     * Starts the Kafka-based catch up of the given {@link ProjectionRepository}.
+     *
+     * <p>Call this method on the server start up for all the Projection Repositories or when
+     * setting up a repository of a new projection type.
+     *
+     * <p>This method starts a
+     * <a href="https://kafka.apache.org/documentation/streams/">Kafka Stream topology</a>.
+     * The source topic with name {@link Topic#eventTopic() spine.core.Event} should exist before
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/catchup/KafkaCatchUp.java]: `should` or `must`? And what if there is no topology — please describe.

File: kafka-storage/src/main/java/io/spine/server/catchup/KafkaCatchUp.java [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+     * required. The newly created topology will start processing from the first event found in
+     * {@code spine.core.Event} topic.
+     *
+     * <p>The passed {@code streamConfig} should not contain {@code application.id} attribute, as
+     * it's assigned automatically depending on the repository type.
+     *
+     * <p>Note: it may take a few seconds before the Kafka Streams topology starts serving. Do not
+     * rely on an immediate start.
+     *
+     * <p>Note: it is required to add {@linkplain #dispatcher Kafka event dispatcher} to
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/catchup/KafkaCatchUp.java]: Having two `Note:` totally destroys the purpose. I'd remove both — it's going to be OK with no `Note:` in both cases.

File: kafka-storage/src/main/java/io/spine/server/catchup/KafkaCatchUp.java [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+        return voidInstance();
+    }
+
+    private static String repositoryKey(ProjectionRepository<?, ?, ?> repository) {
+        final String typeName = repository.getEntityStateType().getTypeName();
+        return typeName;
+    }
+
+    /**
+     * @return an instance of {@link Void} class at compile time and {@code null} at runtime.
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/catchup/KafkaCatchUp.java]: I don't understand. Please put that into some context. 

Also, if this returns `null`, I'd mark it somehow. Remember, we don't return `null`s anywhere by default.

File: kafka-storage/src/main/java/io/spine/server/storage/kafka/KafkaWrapper.java [hunk 8]
```diff
@@ -86,6 +93,62 @@
     private final Lock lock = new ReentrantLock();
 
     /**
+     * Creates a new instance of {@code KafkaWrapper}.
+     *
+     * <p>This method is an overload of
+     * {@link #create(Properties, Properties, Consistency, Duration)} with the recommended values
+     * of the missing arguments. A call to this method with parameters {@code producerConfig} and
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/storage/kafka/KafkaWrapper.java]: `omitted`. `Missing` is usually accidental. `Omitted` is on purpose.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

File: deployment/local-kafka/src/main/resources/config/kafka-streams.properties [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+#
+# This file contains the Kafka Streams configuration for the Kafka-based catch up.
+#
+# Note that the `application.id` config is absent, since it is auto-generated by the library when
```
Comment (armiol) [inline deployment/local-kafka/src/main/resources/config/kafka-streams.properties]: Before this sentence, ~80% of users had no idea there could be `application.id` property. Either there should be some reference (such as a link to Kafka docs, which would explain the property); or this sentence should be rewritten.

File: kafka-storage/src/main/java/io/spine/server/catchup/KafkaCatchUp.java [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+ *
+ * <p>The general flow of the catch up is following:
+ * <ol>
+ *     <li>An event is dispatched by the {@link io.spine.server.event.EventBus EventBus} into
+ *     the {@linkplain #dispatcher(Collection, KafkaWrapper) Kafka event dispatcher}.
+ *     <li>The dispatcher publishes the event into a dedicated Kafka {@linkplain Topic topic} with
+ *     name {@code spine.core.Event}.
+ *     <li>All the events are grouped by the target {@code ProjectionRepository} and dispatched
+ *     into the repository synchronously.
+ *     <li>The repository handles the events by its own.
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/catchup/KafkaCatchUp.java]: `by itself`? Or `on its own`, which has a slightly different meaning.

File: kafka-storage/src/main/java/io/spine/server/catchup/KafkaCatchUp.java [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+        return voidInstance();
+    }
+
+    private static String repositoryKey(ProjectionRepository<?, ?, ?> repository) {
+        final String typeName = repository.getEntityStateType().getTypeName();
+        return typeName;
+    }
+
+    /**
+     * Returns {@code null}; performs to action.
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/catchup/KafkaCatchUp.java]: To what action?

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

**Pull request merged**